### PR TITLE
[GPU/OpenCL] Initial version of Transpose function with OpenCL ops @open sesame 10/04 17:03

### PR DESCRIPTION
**Pull request by niket-agarwal**
Added initial version of Transpose function for GPU. This is a basic implementation using naive kernel.

**Changes added with this PR:**

Incorporated kernels for GPU to transpose about different axes.
Added `unittest_layers_transpose_cl.cpp` to test Transpose function on GPU.

Signed-off-by: Niket Agarwal [niket.a@samsung.com](mailto:niket.a@samsung.com)

### CODE REVIEW

**Review by EunjuYang — COMMENTED** [review]
Why do you implement a transpose _layer_ instead of  tensor operation? It seems not consistent with the CPU-side implementation.
In addition, this transpose layer only supports transposing N, C axis only, which is not generalized.

**Comment by skykongkong8** [comment]
> Why do you implement a transpose _layer_ instead of tensor operation? It seems not consistent with the CPU-side implementation. In addition, this transpose layer only supports transposing N, C axis only, which is not generalized.

I agree.
To committer, please refer to `Tensor::transpose()` for generalized kernel implementation.

**Comment by niket-agarwal** [comment]
> Why do you implement a transpose _layer_ instead of tensor operation? It seems not consistent with the CPU-side implementation. In addition, this transpose layer only supports transposing N, C axis only, which is not generalized.

Okay I'll write it as a tensor operation and come up with a more generalized implementation.

**Comment by jijoongmoon** [comment]
Overall, it is ok. but please check if you can use permute layer as in https://github.com/nnstreamer/nntrainer/blob/main/nntrainer/layers/permute_layer.cpp

File: nntrainer/layers/cl_layers/transpose_cl.cpp [hunk 1]
```diff
@@ -0,0 +1,357 @@
...
+    // Calculate n, c, h, w from the global ids
+    int h = get_global_id(0);
+    int w = get_global_id(1);
+
+    if (h < H && w < W) {
+        for (int c = 0; c < C; ++c) {
+            for (int n = 0; n < N; ++n) {
+                // Calculate the input and output indices
+                int input_index = n * (C * input_size) + c * input_size + h * W + w;
+                int output_index = c * (N * output_size) + n * output_size + h * W + w;
```
Comment (djeong20) [inline nntrainer/layers/cl_layers/transpose_cl.cpp:43]: does the current kernel support both formats, channels last and channels first, or designed for channels first only?

File: nntrainer/layers/cl_layers/transpose_cl.cpp [hunk 2]
```diff
@@ -0,0 +1,357 @@
...
+#include <iostream>
+#include <layer_context.h>
+#include <nntrainer_error.h>
+#include <nntrainer_log.h>
+#include <node_exporter.h>
+
+std::string transpose_cl_kernel_fp16_ =
+  R"(
+    #pragma OPENCL EXTENSION cl_khr_fp16 : enable
+    __kernel void transpose_cl_fp16(__global const half* in,
```
Comment (SeoHyungjun) [inline nntrainer/layers/cl_layers/transpose_cl.cpp:24]: If you look at other functions, they don't use n, c, h, w; instead, they use accurate names. How about modifying this function as below?
I believe it is a good idea to use similar naming design even though this code is a GPU kernel code.


For example, like the function below:
```cpp
void TransposeLayerCl::transpose_cl(const float *in, float *res,
                                    unsigned int input_batch_size,
                                    unsigned int input_channels,
                                    unsigned int input_height,
                                    unsigned int input_width,
                                    RunLayerContext &context)
```

**Review by SeoHyungjun — APPROVED** [review]

File: nntrainer/layers/cl_layers/transpose_cl.cpp (hunk 1 above)
Comment (niket-agarwal) [reply nntrainer/layers/cl_layers/transpose_cl.cpp]: Only channels first. (nchw)

File: nntrainer/layers/cl_layers/transpose_cl.h [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+
+  /**
+   * @copydoc Layer::calcDerivative(RunLayerContext &context)
+   */
+  void calcDerivative(RunLayerContext &context) override;
+
+  /**
+   * @copydoc bool supportBackwarding() const
+   */
+  bool supportBackwarding() const override { return true; };
```
Comment (djeong20) [inline nntrainer/layers/cl_layers/transpose_cl.h:73]: I believe this should be false for now.

```suggestion
  bool supportBackwarding() const override { return false; };
```

**Review by djeong20 — COMMENTED** [review]
Overall, awesome work! However, I think there's room to improve this patch 💯 
please check the comments.

+) Please rebase as well

File: nntrainer/layers/cl_layers/transpose_cl.cpp [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+    }
+  }
+
+  context.setOutputDimensions(dim);
+}
+
+void TransposeLayerCl::forwarding(RunLayerContext &context, bool training) {
+  Tensor &in = context.getInput(SINGLE_INOUT_IDX);
+  Tensor &out = context.getOutput(SINGLE_INOUT_IDX);
+  Transpose_i_cl("1:0:2", in, out, context);
```
[the remark this answers is not in the document]
Comment (niket-agarwal) [reply nntrainer/layers/cl_layers/transpose_cl.cpp:45]: Took CPU implementation as reference and only this specific direction was being called.

File: nntrainer/tensor/cl_operations/blas_kernel_interface.cpp [hunk 5]
```diff
@@ -244,4 +244,58 @@ void add_i_cl(Tensor const &input, Tensor &result, RunLayerContext &context) {
...
+    // for transpose about height and width
+    else if (direction[0] == '0' && direction[2] == '2') {
+      transpose_cl_axis1(data, rdata, input_batch_size, input_channels,
+                         input_height, input_width, context);
+    }
+    // for transpose about channels and width
+    else if (direction[0] == '2' && direction[2] == '1') {
+      transpose_cl_axis2(data, rdata, input_batch_size, input_channels,
+                         input_height, input_width, context);
+    }
```
[the remark this answers is not in the document]
Comment (niket-agarwal) [reply nntrainer/tensor/cl_operations/blas_kernel_interface.cpp:274]: Function here is used to transpose about two axes. Others can be done in a two-step call.
 0:1:2 ----> 0:2:1 ---->1:2:0
 0:1:2 ----> 1:0:2 ----> 2:0:1

File: nntrainer/tensor/cl_operations/blas_kernels.cpp [hunk 6]
```diff
@@ -119,6 +119,87 @@ std::string sscal_cl_kernel_ =
...
+                int input_index = n * (channels * height * width) + c * (height * width) + h * width + w;
+                int output_index = n * (channels * height * width) + w * (height * channels) + h * channels + c;
+
+                // Transpose channel and width, copying data from input to output
+                output[output_index] = in[input_index];
+            }
+        }
+    }
+
+})";
```
[the remark this answers is not in the document]
Comment (niket-agarwal) [reply nntrainer/tensor/cl_operations/blas_kernels.cpp:201]: Even we are using a single Transpose_i_cl regardless of the permutation of the dimensions. All these different kernel calls would be taken care internally based on the direction.

File: test/unittest/layers/unittest_layers_transpose_cl.cpp [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+//                              LayerGoldenTestParamOptions::USE_INC_FORWARD,
+//                            "nchw", "fp32", "fp32");
+
+// auto transpose_basic_plain_axis2 =
+//   LayerGoldenTestParamType(nntrainer::createLayer<nntrainer::TransposeLayerCl>,
+//                            {}, "2:3:3:3", "transpose_axis2.nnlayergolden",
+//                            LayerGoldenTestParamOptions::SKIP_CALC_DERIV |
+//                              LayerGoldenTestParamOptions::SKIP_CALC_GRAD |
+//                              LayerGoldenTestParamOptions::USE_INC_FORWARD,
+//                            "nchw", "fp32", "fp32");
```
[the remark this answers is not in the document]
Comment (niket-agarwal) [reply test/unittest/layers/unittest_layers_transpose_cl.cpp:49]: Only a specific direction transpose call (1:0:2) is made currently in reference to the CPU implementation. Hence commented out the other testcases.

**Comment by baek2sm** [comment]
Nice work! But.. As you already know, this transpose layer is not a general transpose layer but rather one specifically tailored for the limited environment of the llama application. As others have already mentioned in their comments, generalization would require using either the tensor::transpose or permute layers rather than the transpose layer within the llama application which only works under specific settings.

**Comment by niket-agarwal** [comment]
> Nice work! But.. As you already know, this transpose layer is not a general transpose layer but rather one specifically tailored for the limited environment of the llama application. As others have already mentioned in their comments, generalization would require using either the tensor::transpose or permute layers rather than the transpose layer within the llama application which only works under specific settings.

Whatever we have done here is also a generalized implementation only. Using transpose_i_cl which is generalised internally and can handle all the directions and would work under all settings.

**Review by djeong20 — COMMENTED** [review]
First of all, thank you for the pull request and the hard work! 
I think what @baek2sm meant is the generalization of a layer, not the kernel. I agree that `transpose_i_cl()` is generalized internally and can handle transpose two axes and this is a good starting point 👍. However, to explain, TransposeLayer was added only for the LLaMA application and is not used for all purposes. (we use Tensor::transpose() and PermuteLyaer instead). It is currently not used in the LLaMA application as well.
In my opinion, all works under nntrainer/tensor/cl_operations are good to go (nice work!). While `TransposeLayer` is not used, I think `PermuteLayerCl` would be a better option (maybe this could be the next PR I guess).

**Review by EunjuYang — APPROVED** [review]
Overall, LGTM! Except for the clang format :) Please check the format.

(+ Additional comment)
In MHA layer, transpose is required.
I agree with that it would be better to implement the transpose feature as a tensor operation. For this time, however, it is okay to merge this PR to enable GPU kernels to run LLM.

**Pull request closed without merging**